internal/archive/archive.go

```diff
@@ -178,20 +184,44 @@ func openUbuntu(options *Options) (Archive, error) {
 
 func (index *ubuntuIndex) fetchRelease() error {
 	logf("Fetching %s %s %s suite details...", index.label, index.version, index.suite)
-	reader, err := index.fetch("Release", "", fetchDefault)
+	reader, err := index.fetch("InRelease", "", fetchDefault)
+	if err != nil {
+		return err
+	}
+	defer reader.Close()
+
+	data, err := io.ReadAll(reader)
 	if err != nil {
 		return err
 	}
 
-	ctrl, err := control.ParseReader("Label", reader)
+	// Decode the signature(s) and verify the InRelease file.
+	// The InRelease file may have multiple signatures from
+	// different keys. Verify that at least one signature is valid
+	// against the archive's set of public keys. Unlike ``gpg --verify``
+	// which ensures the verification of all signatures, this is in line
+	// with what apt does internally [1].
+	// [1]  https://salsa.debian.org/apt-team/apt/-/blob/4e344a4c1d2862b7cdb900a20222bc22ac5edcf7/methods/gpgv.cc#L553-557
+	sigs, body, content, err := setup.DecodeClearSigned(data)
+	if err != nil {
+		return fmt.Errorf("corrupted archive InRelease file: invalid signature")
+	}
+	err = setup.VerifyAnySignature(index.archive.publicKeys, sigs, body)
+	if err != nil {
+		return fmt.Errorf("cannot verify signature in the InRelease file")
+	}
+
+	// Using ``content`` here because ``body`` has CRLF endings.
+	// TODO	figure out how to use either ``body`` or ``content``.
+	ctrl, err := control.ParseString("Label", string(content))
```

Comment:
The problem is per GPG's RFC ([source](https://www.rfc-editor.org/rfc/rfc4880#section-5.2.4)):
>   For binary document signatures (type 0x00), the document data is
>   hashed directly.  For text document signatures (type 0x01), the
>   document is canonicalized by converting line endings to \<CR\>\<LF\>,
>   and the resulting data is hashed.

That means that the line endings of `body` are changed to CRLF, and we need `body` to verify the signature above. We also need `content` because the control module is designed to parse control files whose line endings are expected to be `\n`.

The only solution would be using only `body` (we need the CRLF endings for verifying the signature) and then adding some parsing logic to contemplate the different line endings. However, it has the following issues:
1. If we change the control parser we are accepting CRLF endings which are actually invalid in control files.
2. If we add another parser then we have essentially duplicated our parsing logic, which means they can deviate.

If you think it is worth it to remove the usage of `body` I would advocate for 2. and test the new implementation with the same tests as the existing one.

Reply:
Understood, but the shape of the API on our side is still not great: it has two copies of the same data, with names that make it sound different in unclear ways, but that it's actually just the exact same data with different line endings. Worse, we don't even know what the actual original content looked like, I suspect, because we don't have two copies of that data in the actual archive.

So my suggestion is to return one single copy of the data, which is the real one in canonical form and can have the signature validated, and then just replace the line endings before parsing.

Reply:
I have followed your suggestion in https://github.com/canonical/chisel/pull/106/commits/96a4ad62c149e045795d7215e91609516f28ee4a.

I agree it makes the code a lot more readable. However, performance-wise, it is a bad as having two copies of body because we have three now. The inner decoder returns two copies (we ignore one but it takes one GC pass to clean it up) and we are creating another copy by replacing the line endings.

Reply:
Rafid had a very good point about what happens if the InRelease format we download has \<CR\>\<LF\> endings. By changing it manually we would not know if that was the case or not.

Fortunately, apt recognizes both ([ref](https://salsa.debian.org/apt-team/apt/-/commit/dcc0759fea4bf65d5477678e287880927d2cc9be)) which means we should as well.